File: unpruuf/app/src/main/java/com/nexonai/unpruuf/screens/lock/LockViewModel.kt
package com.nexonai.unpruuf.screens.lock

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.nexonai.unpruuf.data.db.ContactDao
import com.nexonai.unpruuf.data.repository.InMemoryMessageStore
import com.nexonai.unpruuf.domain.network.IdentityManager
import com.nexonai.unpruuf.domain.network.ratchet.RatchetSessionManager
import com.nexonai.unpruuf.domain.security.AppLockManager
import com.nexonai.unpruuf.domain.security.PinManager
import com.nexonai.unpruuf.domain.security.PinResult
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject

@HiltViewModel
class LockViewModel @Inject constructor(
    private val pinManager: PinManager,
    private val appLockManager: AppLockManager,
    private val contactDao: ContactDao,
    private val messageStore: InMemoryMessageStore,
    private val identityManager: IdentityManager,
    private val ratchetSessionManager: RatchetSessionManager
) : ViewModel() {

    val isPinSet: Boolean get() = pinManager.isPinSet()
    val isBiometricEnabled: Boolean get() = pinManager.isBiometricEnabled()

    private val _error = MutableStateFlow<String?>(null)
    val error = _error.asStateFlow()

    /** Ersteinrichtung: normale PIN + Panik-PIN festlegen. */
    fun setupPins(pin: String, panicPin: String): Boolean {
        if (pin.length < 4) { _error.value = "PIN must be at least 4 digits."; return false }
        if (pin == panicPin) { _error.value = "Panic PIN must differ from the PIN."; return false }
        pinManager.setPins(pin, panicPin)
        appLockManager.unlock()
        return true
    }

    /** Entsperren. Panik-PIN löscht alle Kontakte und öffnet die App leer. */
    fun submitPin(input: String) {
        when (pinManager.check(input)) {
            PinResult.NORMAL -> {
                _error.value = null
                appLockManager.unlock()
            }
            PinResult.PANIC -> {
                viewModelScope.launch {
                    wipeEverything()
                    _error.value = null
                    appLockManager.unlock()
                }
            }
            PinResult.WRONG -> {
                _error.value = "Wrong PIN."
            }
        }
    }

    /** Called after a successful [androidx.biometric.BiometricPrompt] authentication — same
     *  success path as [PinResult.NORMAL], without checking any PIN, since the biometric check
     *  itself is the proof. Never usable for the panic path: there is no biometric equivalent
     *  of the panic PIN, so a fingerprint can never wipe/open the app empty. */
    fun biometricUnlock() {
        _error.value = null
        appLockManager.unlock()
    }

    fun clearError() { _error.value = null }

    private suspend fun wipeEverything() {
        messageStore.zeroizeAll()
        withContext(Dispatchers.IO) {
            contactDao.getAllContactsOnce().forEach {
                contactDao.deleteById(it.id)
                ratchetSessionManager.deleteSession(it.id)
            }
            identityManager.getAllContactOnionKeys().keys.forEach {
                identityManager.removeContactOnionPrivKey(it)
            }
        }
    }
}

File: unpruuf/ios/UnpruufApp/Sources/ViewModels/LockViewModel.swift
import Combine
import Foundation

/// Port of the Android app's `LockViewModel.kt` — PIN setup/check, panic-PIN wipe, biometric
/// unlock. Same DI shape as this app's other view models (plain `init(env:)`, no framework).
@MainActor
final class LockViewModel: ObservableObject {
    @Published var errorMessage: String?

    private let env: AppEnvironment
    private var pinManager: PinManager { env.pinManager }
    private var appLockManager: AppLockManager { env.appLockManager }

    init(env: AppEnvironment) {
        self.env = env
    }

    var isPinSet: Bool { pinManager.isPinSet }
    var isBiometricEnabled: Bool { pinManager.isBiometricEnabled }

    /// First-time setup: normal PIN + a separate panic PIN.
    func setupPins(pin: String, panicPin: String) {
        guard pin.count >= 4 else { errorMessage = "PIN must be at least 4 digits."; return }
        guard pin != panicPin else { errorMessage = "Panic PIN must differ from the PIN."; return }
        pinManager.setPins(pin: pin, panicPin: panicPin)
        appLockManager.unlock()
    }

    /// Checks a typed PIN. The panic PIN instantly wipes contacts + messages and opens the app
    /// empty — see `AppEnvironment.wipeContactsAndMessages()` for exactly what that does and,
    /// just as importantly, what it deliberately leaves untouched.
    func submitPin(_ input: String) {
        switch pinManager.check(input) {
        case .normal:
            errorMessage = nil
            appLockManager.unlock()
        case .panic:
            env.wipeContactsAndMessages()
            errorMessage = nil
            appLockManager.unlock()
        case .wrong:
            errorMessage = "Wrong PIN."
        }
    }

    /// Called after a successful `LAContext` biometric evaluation — same success path as
    /// `.normal`, without checking any PIN, since the biometric check itself is the proof. Never
    /// usable for the panic path: there is no biometric equivalent of the panic PIN.
    func biometricUnlock() {
        errorMessage = nil
        appLockManager.unlock()
    }

    func clearError() { errorMessage = nil }
}
